File: src/units.cpp
#include "units.hpp"

static float convertMeterToUnit(float value, Unit to) {
  switch (to) {
    case Unit::Centimeter:
      return value * 100;
    case Unit::Millimeter:
      return value * 1000;
    case Unit::Inch:
      return value * 39.3701;
    case Unit::Foot:
      return value * 3.28084;
    default:
      return value;
  }
}

static float convertUnitToMeter(float value, Unit from) {
  switch (from) {
    case Unit::Centimeter:
      return value / 100;
    case Unit::Millimeter:
      return value / 1000;
    case Unit::Inch:
      return value / 39.3701;
    case Unit::Foot:
      return value / 3.28084;
    default:
      return value;
  }
}

float convertUnit(float value, Unit from, Unit to) {
  float meterValue = convertUnitToMeter(value, from);
  return convertMeterToUnit(meterValue, to);
}

String getUnitSymbol(Unit unit) {
  switch (unit) {
    case Unit::Meter:
      return "m";
    case Unit::Centimeter:
      return "cm";
    case Unit::Millimeter:
      return "mm";
    case Unit::Inch:
      return "in";
    case Unit::Foot:
      return "ft";
    default:
      return "";
  }
}

Unit getUnitForSymbol(String symbol) {
  if (symbol == "m") return Unit::Meter;
  if (symbol == "cm") return Unit::Centimeter;
  if (symbol == "mm") return Unit::Millimeter;
  if (symbol == "in") return Unit::Inch;
  if (symbol == "ft") return Unit::Foot;
  return Unit::Centimeter;
}

File: dashboard/src/lib/units.ts
/**
 * Utility enum class for units of measurement.
 */
export class Unit {
  static readonly METER = new Unit("Meter", "m", 1, 3);
  static readonly CENTIMETER = new Unit("Centimeter", "cm", 100, 1);
  static readonly MILLIMETER = new Unit("Millimeter", "mm", 1000, 0);
  static readonly INCH = new Unit("Inch", "in", 39.3701, 2);
  static readonly FOOT = new Unit("Foot", "ft", 3.28084, 3);

  private constructor(public readonly name: string, public readonly symbol: string, public readonly conversionFactor: number, public readonly displayDecimals: number) {}

  /**
   * Get a unit from a symbol.
   *
   * @param symbol - The symbol to get the unit for.
   * @returns The unit
   */
  public static fromSymbol(symbol: "m" | "cm" | "mm" | "in" | "ft"): Unit {
    return Object.values(Unit).find(u => u.symbol === symbol);
  }

  /**
   * Helper method to convert a value from this unit to another.
   *
   * @param to - The unit to convert to.
   * @param value - The value to convert.
   * @returns The converted value.
   */
  public convertTo(to: Unit, value: number): number {
    return convertUnit(value, this, to);
  }

  /**
   * Helper method to convert a value from another unit to this unit.
   *
   * @param from - The unit to convert from.
   * @param value - The value to convert.
   * @returns The converted value.
   */
  public convertFrom(from: Unit, value: number): number {
    return convertUnit(value, from, this);
  }

  public toString(): string {
    return this.symbol;
  }
}

/**
 * Convert a value from one unit to another.
 *
 * @param value - The value to convert.
 * @param from - The unit to convert from.
 * @param to - The unit to convert to.
 * @returns The converted value.
 */
export function convertUnit(value: number, from: Unit, to: Unit): number {
  return value * to.conversionFactor / from.conversionFactor;
}
